### PR TITLE
Simple public source page

### PR DESCRIPTION
**Pull request by antoine-le-calloch**
The purpose of this issue is to make a private source page public by generating simple html that will be added to the cache and made available to the public. And in addition to display a public page with all the souces and versions of this source currently in a public mode. 

## Version 1
- Creation of a table to store whether a source page is public or private and what informations are display
- A public source page is generated when a page is made public, and is added to the cache
- The source page is fetched from the cache on access
- Display a good public page with source data on it
- Display photometry on public page
- Allow to download photometry in csv
- Display classification on public page 
- Search button on public sources page to filter all the sources
- Add a dialogue on private source page to manage access information, display actual public source versions and publish a new version
- Access information allow to include or not photometry, classification and to manage access by group and stream
- Versions are manage by date in UTC
- Display the time ago a source has been publish on public sources page

### CODE REVIEW

**Comment by mcoughlin** [comment]
@antoine-le-calloch we should put the proposed component reorganization in its own PR. But good for @Theodlz to see it for sure.

**Comment by antoine-le-calloch** [comment]
@Theodlz here is the pull request for the public source and sources page.

**Comment by antoine-le-calloch** [comment]
> @antoine-le-calloch we should put the proposed component reorganization in its own PR. But good for @Theodlz to see it for sure.

Yes indeed but complicated to separate the two now.

**Comment by Theodlz** [comment]
It's fine let's just keep both here, but keep that in mind for later. When you can, splitting PRs and keeping them self contained makes it much easier for us to review and hence merge it quickly @antoine-le-calloch

File: skyportal/handlers/public/source_page.py
Comment (Theodlz) [inline skyportal/handlers/public/source_page.py:42]: Usually for API handlers that return multiple things as a list, or just one (if ID is provided), we have a specific dosctring syntax with the "single" and "multiple" keywords. Please correct that here before merging.

File: skyportal/handlers/public/source_page.py
Comment (Theodlz) [inline skyportal/handlers/public/source_page.py:73]: This is an outdated and incorrect syntax. Please just use the DBSession() as usual.

File: skyportal/model_util.py
Comment (Theodlz) [inline skyportal/model_util.py:14]: Maybe manage sources is enough? A question for @mcoughlin. That or the upload data permission idk.

File: skyportal/models/public_pages/public_source_page.py
Comment (Theodlz) [inline skyportal/models/public_pages/public_source_page.py:5]: Hmmm you shouldn't need ligo skymap to plot anything here??? That's for GCN events stuff (hence the ligo)

File: skyportal/models/public_pages/public_source_page.py
Comment (Theodlz) [inline skyportal/models/public_pages/public_source_page.py:1]: The model file is very nice overall, nice job!

File: skyportal/tests/frontend/test_user_management.py
Comment (Theodlz) [inline skyportal/tests/frontend/test_user_management.py:66]: Not sure why you need the by.xpath, that's already exactly what our custom driver offers. The `wait_for_xpath()` should return the element. Please remove that imports and use the existing syntax if possible

File: skyportal/tests/frontend/test_user_management.py
Comment (Theodlz) [inline skyportal/tests/frontend/test_user_management.py:67]: same as the scroll to element and click, I believe we might even have a simple scroll to element. Please check the driver class and use that if available, if not then you can keep this of course, it's completely fine.

File: skyportal/utils/cache.py
Comment (Theodlz) [inline skyportal/utils/cache.py:114]: Nice!

File: skyportal/utils/cache.py
Comment (Theodlz) [inline skyportal/utils/cache.py:128]: I would have a try except here maybe, just to avoid an edge case where it gets deleted exactly between when you run your if statement above and when you delete it, or if for some reason we don't have delete rights.

File: skyportal/models/gcn.py
Comment (Theodlz) [inline skyportal/models/gcn.py:53]: In a following PR, we'll want to move the GCN public pages to the public_pages folder too. Just something to keep in mind

File: static/js/components/NewsFeed.jsx
Comment (Theodlz) [inline static/js/components/NewsFeed.jsx:7]: Ah can you actually change that one import of multiple elements to 3 distinct direct imports like we use for everything else? These can increase the webpack size, and looks like we missed that one!

File: static/js/components/NotificationSettingsSelect.jsx
Comment (Theodlz) [inline static/js/components/NotificationSettingsSelect.jsx:18]: Same comment as before, please split in 4 distinct imports for webpack splitting

File: static/js/components/Notifications.jsx
Comment (Theodlz) [inline static/js/components/Notifications.jsx:134]: Just making sure that doesn't break any tests or features

File: static/js/components/SelectWithChips.jsx
Comment (Theodlz) [inline static/js/components/SelectWithChips.jsx:10]: Same comment as before, please split in distinct imports for webpack splitting

File: static/js/components/ShiftSummary.jsx
Comment (Theodlz) [inline static/js/components/ShiftSummary.jsx:15]: Same comment as before, please split in 4 distinct imports for webpack splitting

File: static/js/components/ShiftSummary.jsx
Comment (Theodlz) [inline static/js/components/ShiftSummary.jsx:19]: Same comment as before, please split in 4 distinct imports for webpack splitting

File: static/js/components/ThumbnailList.jsx
Comment (Theodlz) [inline static/js/components/ThumbnailList.jsx:1]: Just making sure this is normal, hard to tell with the git UI if that went somewhere else

File: static/js/components/photometry/PhotometryMagsys.jsx
Comment (Theodlz) [inline static/js/components/photometry/PhotometryMagsys.jsx:84]: Nice catch. Thanks!

File: static/js/components/source/SourceCoordinates.jsx
Comment (Theodlz) [inline static/js/components/source/SourceCoordinates.jsx:1]: Very weird order for those imports, please fix those and lint!

File: static/js/components/source/source_publish/SourcePublish.jsx
Comment (Theodlz) [inline static/js/components/source/source_publish/SourcePublish.jsx:8]: Import order to verify maybe? Usually all the MUI stuff comes after the react stuff.

File: static/js/components/source/source_publish/SourcePublish.jsx
Comment (Theodlz) [inline static/js/components/source/source_publish/SourcePublish.jsx:177]: maybe  "The data you selected" rather than info?

File: static/js/components/source/source_publish/SourcePublishHistory.jsx
Comment (Theodlz) [inline static/js/components/source/source_publish/SourcePublishHistory.jsx:72]: the code speaks for itself here, you can remove the comment

File: static/js/utils/csv.js
Comment (Theodlz) [inline static/js/utils/csv.js:1]: nice

File: static/js/utils/photometry.js
Comment (Theodlz) [inline static/js/utils/photometry.js:3]: Thanks a lot for refactoring these!

File: static/js/utils/photometry.js
Comment (Theodlz) [inline static/js/utils/photometry.js:6]: as discussed though, this was if these 3 filters are the only one used, and it's an incomplete list. Please add the other filters. You can copy/paste the list returned by the configHandler if it need to be static

File: static/js/utils/photometry.js
Comment (Theodlz) [inline static/js/utils/photometry.js:9]: That will avoid having every non-ztf blue, which is clearly a problem and blocker before we can use these public pages

**Review by Theodlz — COMMENTED** [review]
First preliminary review, I looked at all 280 files changed (yikes!). I'll let you respond to the review and address comments. When you are done, a couple of things for you to do:
1. Uncomment the lines that are commented out in the webpack config file. I'm talking about the ones that allow you to generate a report (that opens in a page of the browser) of what your webpack is made of, and how it is split. Looks for any anomalies that splitting the files as you did might have created, and post a screenshot of the whole thing here.
2. Run the app in make_production, then compile the webpack manually (the comment can be found in the Dockerfile). Once compiled in your terminal it will show you a bunch of info about the number of chunks, their size, ... please screenshot it (or copy/paste) and put it here. Again, just want to see the potential of your refactor on the webpack. We want to make sure it's not gonna affect anything.

File: .gitignore
Comment (antoine-le-calloch) [reply .gitignore:30]: :white_check_mark: Done.

File: config.yaml.defaults
Comment (antoine-le-calloch) [reply config.yaml.defaults]: ✅ Done.

File: skyportal/handlers/public/source_page.py
Comment (antoine-le-calloch) [reply skyportal/handlers/public/source_page.py:42]: ✅ Done.

File: skyportal/handlers/public/source_page.py
Comment (antoine-le-calloch) [reply skyportal/handlers/public/source_page.py:73]: ✅ Done.

File: skyportal/models/public_pages/public_source_page.py
Comment (antoine-le-calloch) [reply skyportal/models/public_pages/public_source_page.py:5]: ✅ Done.

File: static/js/components/NewsFeed.jsx
Comment (antoine-le-calloch) [reply static/js/components/NewsFeed.jsx:7]: ✅ Done.

File: static/js/components/NotificationSettingsSelect.jsx
Comment (antoine-le-calloch) [reply static/js/components/NotificationSettingsSelect.jsx:18]: ✅ Done.

File: static/js/components/SelectWithChips.jsx
Comment (antoine-le-calloch) [reply static/js/components/SelectWithChips.jsx:10]: ✅ Done.

File: static/js/components/ShiftSummary.jsx
Comment (antoine-le-calloch) [reply static/js/components/ShiftSummary.jsx:15]: ✅ Done.

File: static/js/components/ShiftSummary.jsx
Comment (antoine-le-calloch) [reply static/js/components/ShiftSummary.jsx:19]: ✅ Done.

File: static/js/components/source/SourceCoordinates.jsx
Comment (antoine-le-calloch) [reply static/js/components/source/SourceCoordinates.jsx:1]: ✅ Done.

File: static/js/components/source/source_publish/SourcePublish.jsx
Comment (antoine-le-calloch) [reply static/js/components/source/source_publish/SourcePublish.jsx:8]: ✅ Done.

File: static/js/components/source/source_publish/SourcePublish.jsx
Comment (antoine-le-calloch) [reply static/js/components/source/source_publish/SourcePublish.jsx:177]: ✅ Done.

File: static/js/components/source/source_publish/SourcePublishHistory.jsx
Comment (antoine-le-calloch) [reply static/js/components/source/source_publish/SourcePublishHistory.jsx:72]: ✅ Done.

File: skyportal/utils/cache.py
Comment (antoine-le-calloch) [reply skyportal/utils/cache.py:128]: ✅ Done.

File: static/js/components/ShiftSummary.jsx
Comment (Theodlz) [inline static/js/components/ShiftSummary.jsx:15]: The import order is incorrect. All the MUI or npm installable stuff should come before the baselayer and skyportal imports. Just like in python

Comment (Theodlz) [reply static/js/components/ShiftSummary.jsx:15]: Maybe it was already incorrect, in which case please fix it :)

**Review by Theodlz — CHANGES_REQUESTED** [review]
Super small comment, otherwise this is great. I will run the code today and suggest any modifications of the feature itself if need be!

File: skyportal/model_util.py
Comment (mcoughlin) [reply skyportal/model_util.py:14]: Maybe we stick with Manage Sources for now, and reconsider after seeing how it works in production.

Comment (Theodlz) [reply skyportal/model_util.py:14]: @antoine-le-calloch tagging you so you see that. You can remove that new permission and use Manage sources everywhere you used `Manage sources publishing`

**Comment by antoine-le-calloch** [comment]
> First preliminary review, I looked at all 280 files changed (yikes!). I'll let you respond to the review and address comments. When you are done, a couple of things for you to do:
> 
>     1. Uncomment the lines that are commented out in the webpack config file. I'm talking about the ones that allow you to generate a report (that opens in a page of the browser) of what your webpack is made of, and how it is split. Looks for any anomalies that splitting the files as you did might have created, and post a screenshot of the whole thing here.

Here is the first report for the main branch and for this PR.
[skyportal_report.zip](https://github.com/skyportal/skyportal/files/15342440/skyportal_report.zip)

**Comment by Theodlz** [comment]
> > First preliminary review, I looked at all 280 files changed (yikes!). I'll let you respond to the review and address comments. When you are done, a couple of things for you to do:
> > ```
> > 1. Uncomment the lines that are commented out in the webpack config file. I'm talking about the ones that allow you to generate a report (that opens in a page of the browser) of what your webpack is made of, and how it is split. Looks for any anomalies that splitting the files as you did might have created, and post a screenshot of the whole thing here.
> > ```
> 
> Here is the first report for the main branch and for this PR. [skyportal_report.zip](https://github.com/skyportal/skyportal/files/15342440/skyportal_report.zip)

The bundles/chunks look very similar, though I notice that after your refactoring and import bugfixes one of the medium-sized node_modules chunks seems to be broken down further, though it looks like the static code bundle is slightly larger (but you added some code and files, so that is not surprising).

The overall size is comparable, and node_modules seem to be better packaged. Great job! We might want to track down any remaining imports that would benefit from some rewriting but this is great. Thanks for your work @antoine-le-calloch!

**Comment by Theodlz** [comment]
@antoine-le-calloch could you please make me a collaborator on your fork? That way I can also fix merge conflicts before re-reviewing when needed.

Please also note that some migrations were merged to main, so we'll need to update the down-revision from your PR's migration script again before merging anything.

File: skyportal/models/public_pages/public_source_page.py
Comment (Theodlz) [inline skyportal/models/public_pages/public_source_page.py:153]: Is that needed??? We already have a `created_at` column on every data model, I find that very surprising.

File: skyportal/models/public_pages/public_source_page.py
Comment (Theodlz) [inline skyportal/models/public_pages/public_source_page.py:52]: Do we need to have a duplicate of `created_at`??? I would strongly advise against it. We already have that in the DB, pretty much in ISO format too.

File: static/js/components/templates/HeaderContent.jsx.template
Comment (Theodlz) [inline static/js/components/templates/HeaderContent.jsx.template:38]: Any reason why this had to be removed?

**Review by Theodlz — COMMENTED** [review]
Very very nice, the frontend is impressively smooth! Very minor comments mostly about the backend.

One thing to change about the logic to filter out photometry: Streams and group restrictions should be an OR statement, not an AND. Maybe we'll want that to be an option later in the future but for now an OR is much more convenient.

File: skyportal/models/public_pages/public_source_page.py
Comment (antoine-le-calloch) [reply skyportal/models/public_pages/public_source_page.py:62]: Yeah you right this option is only a leftover from the old implementation of the public page before the versions were added. Indeed now it is no longer useful. I can implement it but in the current versioning logic I don't see in which case it could be useful. Do you think this is useful ? @Theodlz  @mcoughlin

Comment (Theodlz) [reply skyportal/models/public_pages/public_source_page.py:62]: Not at all you can remove that column (from both the model and the migration script btw), but also as I mentioned we already have that column called `created_at` on every table (comes from the Base class that all models heritage from), so my comment was just that this is a duplicate of something that is already there.

File: skyportal/models/public_pages/public_source_page.py
Comment (antoine-le-calloch) [reply skyportal/models/public_pages/public_source_page.py:129]: Yeah that create an alembic error and i see this possibility in other api import in model file on the project.

File: static/js/components/source/Source.jsx
Comment (Theodlz) [reply static/js/components/source/Source.jsx:300]: source is from the redux right? We usually don't edit redux objects this way

**Review by Theodlz — COMMENTED** [review]
2 minor comments. Once answered/addressed, ready to merge!

File: skyportal/handlers/api/public_pages/public_source_page.py
Comment (Theodlz) [inline skyportal/handlers/api/public_pages/public_source_page.py:19]: Very nice that you thought about sorting the keys, definitely a requirements here!

File: skyportal/handlers/api/public_pages/public_source_page.py
Comment (Theodlz) [inline skyportal/handlers/api/public_pages/public_source_page.py:96]: Just checking: how has the rounding of these values been decided? Based on what is on the frontend?

**Review by Theodlz — COMMENTED** [review]
minor comments, mostly asking some questions to make sure I understand some of the changes

File: skyportal/handlers/api/public_pages/public_source_page.py
Comment (antoine-le-calloch) [reply skyportal/handlers/api/public_pages/public_source_page.py:96]: Yes exactly i follow the private source display round.

**Comment by Theodlz** [comment]
@antoine-le-calloch migrations were fixed on main if you want to merge your branch with it.

**Review by Theodlz — APPROVED** [review]
Super simple comments, otherwise I am happy with this version as a V1!

**Comment by antoine-le-calloch** [comment]
@Theodlz @mcoughlin. Last corrections have been made.

**Comment by Theodlz** [comment]
I'm about to merge your PR @antoine-le-calloch, any chance to resolve the conflicts that are on main super quick? No worries if you can't do it now, just let me know and I'll take care of it.

**Comment by antoine-le-calloch** [comment]
> I'm about to merge your PR @antoine-le-calloch, any chance to resolve the conflicts that are on main super quick? No worries if you can't do it now, just let me know and I'll take care of it.

I'm doing it now !

**Comment by antoine-le-calloch** [comment]
> I'm about to merge your PR @antoine-le-calloch, any chance to resolve the conflicts that are on main super quick? No worries if you can't do it now, just let me know and I'll take care of it.

Done!

**Pull request merged**
